Detect "Use newtype instead of data" for data family instances

e.g.
```haskell
data family F a

data instance F Int = FInt Int
```
we should suggest `newtype instance` instead here.

`ghc` rejects multiple instances:
<details><summary>Example</summary>
<p>

```haskell
blademaster :: /tmp/df » cat Root.hs
{-# LANGUAGE TypeFamilies #-}

module Root where

data family F a

blademaster :: /tmp/df » cat Left.hs
{-# LANGUAGE TypeFamilies #-}

module Left where

import Root

newtype instance F Int = LeftInt Int

blademaster :: /tmp/df » cat Right.hs
{-# LANGUAGE TypeFamilies #-}

module Right where

import Root

newtype instance F Int = RightInt Int

blademaster :: /tmp/df » cat Leaf.hs
module Leaf where

import Left
import Right

blademaster :: /tmp/df » ghc Leaf.hs
[4 of 4] Compiling Leaf             ( Leaf.hs, Leaf.o )

Leaf.hs:1:1: error:
    Conflicting family instance declarations:
      F Int -- Defined in ‘Left’
      F Int -- Defined in ‘Right’
  |
1 | module Leaf where
  | ^

```

</p>
</details>

What is the situation in which this might be unsafe? As long as you want to use both instance you'd have to import them somewhere along the way, which should trigger this error (?)

--- a/src/Hint/NewType.hs
+++ b/src/Hint/NewType.hs
@@ -100,7 +100,7 @@ singleSimpleField (L loc (TyClD ext decl@(DataDecl _ _ _ _ dataDef)))
         Just WarnNewtype
               { newDecl = L loc $ TyClD ext decl {tcdDataDefn = dataDef
                   { dd_ND = NewType
-                  , dd_cons = map (\(L consloc x) -> L consloc $ dropConsBang x) $ dd_cons dataDef
+                  , dd_cons = dropBangs dataDef
                   }}
               , insideType = inType
               }
@@ -109,11 +109,14 @@ singleSimpleField (L loc (InstD ext inst@(DataFamInstD instExt (DataFamInstDecl 
         Just WarnNewtype
           { newDecl = L loc $ InstD ext $ DataFamInstD instExt $ DataFamInstDecl $ HsIB hsibExt famEqn {feqn_rhs = dataDef
                   { dd_ND = NewType
-                  , dd_cons = map (\(L consloc x) -> L consloc $ dropConsBang x) $ dd_cons dataDef
+                  , dd_cons = dropBangs dataDef
                   }}
               , insideType = inType
               }
 singleSimpleField _ = Nothing
+
+dropBangs :: HsDataDefn GhcPs -> [LConDecl GhcPs]
+dropBangs = map (\(L consloc x) -> L consloc $ dropConsBang x) . dd_cons
 
 -- | Checks whether its argument is a \"simple\" data definition (see 'singleSimpleField')
 -- returning the type inside its constructor if it is.

--- a/src/Hint/NewType.hs
+++ b/src/Hint/NewType.hs
@@ -115,13 +115,13 @@ singleSimpleField (L loc (InstD ext inst@(DataFamInstD instExt (DataFamInstDecl 
               }
 singleSimpleField _ = Nothing
 
--- | Checks whether its argument is a \"simple\" data definition (see 'singleSimpleFieldNew')
+-- | Checks whether its argument is a \"simple\" data definition (see 'singleSimpleField')
 -- returning the type inside its constructor if it is.
 simpleHsDataDefn :: HsDataDefn GhcPs -> Maybe (HsType GhcPs)
 simpleHsDataDefn dataDef@(HsDataDefn _ DataType _ _ _ [L _ constructor] _) = simpleCons constructor
 simpleHsDataDefn _ = Nothing
 
--- | Checks whether its argument is a \"simple\" constructor (see criteria in 'singleSimpleFieldNew')
+-- | Checks whether its argument is a \"simple\" constructor (see criteria in 'singleSimpleField')
 -- returning the type inside the constructor if it is. This is needed for strictness analysis.
 simpleCons :: ConDecl GhcPs -> Maybe (HsType GhcPs)
 simpleCons (ConDeclH98 _ _ _ [] context (PrefixCon [L _ inType]) _)

--- a/src/Hint/NewType.hs
+++ b/src/Hint/NewType.hs
@@ -116,7 +116,7 @@ singleSimpleField (L loc (InstD ext inst@(DataFamInstD instExt (DataFamInstDecl 
 singleSimpleField _ = Nothing
 
 dropBangs :: HsDataDefn GhcPs -> [LConDecl GhcPs]
-dropBangs = map (\(L consloc x) -> L consloc $ dropConsBang x) . dd_cons
+dropBangs = map (fmap dropConsBang) . dd_cons
 
 -- | Checks whether its argument is a \"simple\" data definition (see 'singleSimpleField')
 -- returning the type inside its constructor if it is.

--- a/src/Hint/NewType.hs
+++ b/src/Hint/NewType.hs
@@ -27,6 +27,15 @@ data A = A () -- newtype A = A ()
 newtype Foo = Foo Int deriving (Show, Eq) --
 newtype Foo = Foo { getFoo :: Int } deriving (Show, Eq) --
 newtype Foo = Foo Int deriving stock Show
+data instance Foo Int = Bar Bool -- newtype instance Foo Int = Bar Bool
+data instance Foo Int = Bar {field :: Bool} -- newtype instance Foo Int = Bar {field :: Bool}
+data instance Foo Int = Bar {field :: Int#}
+data instance Foo Int = Bar
+data instance Foo Int = Bar {field1 :: Bool, field2 :: ()}
+newtype instance Foo Int = Bar Bool deriving (Show, Eq) --
+newtype instance Foo Int = Bar {field :: Bool} deriving Show --
+newtype instance Foo Int = Bar {field :: Bool} deriving stock Show
+{-# LANGUAGE RankNTypes #-}; data instance Foo Int = forall a. Show a => Foo a
 </TEST>
 -}
 module Hint.NewType (newtypeHint) where
@@ -51,8 +60,14 @@ newtypeHintDecl _ = []
 
 newTypeDerivingStrategiesHintDecl :: LHsDecl GhcPs -> [Idea]
 newTypeDerivingStrategiesHintDecl decl@(L _ (TyClD _ (DataDecl _ _ _ _ dataDef))) =
-    [ignoreNoSuggestion "Use DerivingStrategies" decl | not $ isData dataDef, not $ hasAllStrategies dataDef]
+    [ignoreNoSuggestion "Use DerivingStrategies" decl | shouldSuggestStrategies dataDef]
+newTypeDerivingStrategiesHintDecl decl@(L _ (InstD _ (DataFamInstD _ (DataFamInstDecl (HsIB _ (FamEqn _ _ _ _ _ dataDef)))))) =
+    [ignoreNoSuggestion "Use DerivingStrategies" decl | shouldSuggestStrategies dataDef]
 newTypeDerivingStrategiesHintDecl _ = []
+
+-- | Determine if the given data definition should use deriving strategies.
+shouldSuggestStrategies :: HsDataDefn GhcPs -> Bool
+shouldSuggestStrategies dataDef = all not [isData dataDef, hasAllStrategies dataDef]
 
 hasAllStrategies :: HsDataDefn GhcPs -> Bool
 hasAllStrategies (HsDataDefn _ NewType _ _ _ _ (L _ xs)) = all hasStrategyClause xs
@@ -80,8 +95,8 @@ data WarnNewtype = WarnNewtype
 -- * Single record field constructors get newtyped - @data X = X {getX :: Int}@ -> @newtype X = X {getX :: Int}@
 -- * All other declarations are ignored.
 singleSimpleField :: LHsDecl GhcPs -> Maybe WarnNewtype
-singleSimpleField (L loc (TyClD ext decl@(DataDecl _ _ _ _ dataDef@(HsDataDefn _ DataType _ _ _ [L _ constructor] _))))
-    | Just inType <- simpleCons constructor =
+singleSimpleField (L loc (TyClD ext decl@(DataDecl _ _ _ _ dataDef)))
+    | Just inType <- simpleHsDataDefn dataDef =
         Just WarnNewtype
               { newDecl = L loc $ TyClD ext decl {tcdDataDefn = dataDef
                   { dd_ND = NewType
@@ -89,9 +104,24 @@ singleSimpleField (L loc (TyClD ext decl@(DataDecl _ _ _ _ dataDef@(HsDataDefn _
                   }}
               , insideType = inType
               }
+singleSimpleField (L loc (InstD ext inst@(DataFamInstD instExt (DataFamInstDecl (HsIB hsibExt famEqn@(FamEqn _ _ _ _ _ dataDef))))))
+    | Just inType <- simpleHsDataDefn dataDef =
+        Just WarnNewtype
+          { newDecl = L loc $ InstD ext $ DataFamInstD instExt $ DataFamInstDecl $ HsIB hsibExt famEqn {feqn_rhs = dataDef
+                  { dd_ND = NewType
+                  , dd_cons = map (\(L consloc x) -> L consloc $ dropConsBang x) $ dd_cons dataDef
+                  }}
+              , insideType = inType
+              }
 singleSimpleField _ = Nothing
 
--- | Checks whether its argument is a \"simple constructor\" (see criteria in 'singleSimpleFieldNew')
+-- | Checks whether its argument is a \"simple\" data definition (see 'singleSimpleFieldNew')
+-- returning the type inside its constructor if it is.
+simpleHsDataDefn :: HsDataDefn GhcPs -> Maybe (HsType GhcPs)
+simpleHsDataDefn dataDef@(HsDataDefn _ DataType _ _ _ [L _ constructor] _) = simpleCons constructor
+simpleHsDataDefn _ = Nothing
+
+-- | Checks whether its argument is a \"simple\" constructor (see criteria in 'singleSimpleFieldNew')
 -- returning the type inside the constructor if it is. This is needed for strictness analysis.
 simpleCons :: ConDecl GhcPs -> Maybe (HsType GhcPs)
 simpleCons (ConDeclH98 _ _ _ [] context (PrefixCon [L _ inType]) _)

--- a/src/Hint/NewType.hs
+++ b/src/Hint/NewType.hs
@@ -67,7 +67,7 @@ newTypeDerivingStrategiesHintDecl _ = []
 
 -- | Determine if the given data definition should use deriving strategies.
 shouldSuggestStrategies :: HsDataDefn GhcPs -> Bool
-shouldSuggestStrategies dataDef = all not [isData dataDef, hasAllStrategies dataDef]
+shouldSuggestStrategies dataDef = not (isData dataDef) && not (hasAllStrategies dataDef)
 
 hasAllStrategies :: HsDataDefn GhcPs -> Bool
 hasAllStrategies (HsDataDefn _ NewType _ _ _ _ (L _ xs)) = all hasStrategyClause xs

